kartothek/core/dataset.py

```diff
@@ -381,20 +381,71 @@ def get_indices_as_dataframe(
         elif columns == []:
             return pd.DataFrame(index=self.partitions)
 
-        dfs = []
         columns_to_scan = columns[:]
         if predicates:
             predicate_columns = columns_in_predicates(predicates)
-            # Don't use set logic to preserve order
-            for col in predicate_columns:
-                if col not in columns_to_scan and col in self.indices:
-                    columns_to_scan.append(col)
+            columns_to_scan = sorted(
+                (predicate_columns & self.indices.keys()) | set(columns)
+            )
+
+            dfs = [
```

Comment:
extreme nit: this could be a generator? Not sure if `pd.concat` would handle that efficiently